pkg/securitypolicy/securitypolicyenforcer_rego.go

```diff
@@ -4,67 +4,500 @@
+
+	return policy, nil
+}
+
+type stringArray []string
 
-func createRegoEnforcer(_ SecurityPolicyState, _, _ []oci.Mount) (SecurityPolicyEnforcer, error) {
-	return &RegoEnforcer{}, nil
+func (array stringArray) marshalRego() string {
+	values := make([]string, len(array))
+	for i, value := range array {
+		values[i] = fmt.Sprintf("%q", value)
+	}
+
+	return fmt.Sprintf("[%s]", strings.Join(values, ","))
+}
+
+func writeCommand(builder *strings.Builder, command []string, indent string) error {
+	array := (stringArray(command)).marshalRego()
+	_, err := builder.WriteString(fmt.Sprintf("%s\"command\": %s,\n", indent, array))
+	return err
 }
 
-func (RegoEnforcer) EnforceDeviceMountPolicy(_, _ string) error {
-	return ErrNotImplemented
+func (e EnvRuleConfig) marshalRego() string {
+	return fmt.Sprintf(`{"pattern": "%s", "strategy": "%s"}`, e.Rule, e.Strategy)
 }
 
-func (RegoEnforcer) EnforceDeviceUnmountPolicy(_ string) error {
-	return ErrNotImplemented
+type envRuleArray []EnvRuleConfig
+
+func (array envRuleArray) marshalRego() string {
+	values := make([]string, len(array))
+	for i, env := range array {
+		values[i] = env.marshalRego()
+	}
+
+	return fmt.Sprintf("[%s]", strings.Join(values, ","))
 }
 
-func (RegoEnforcer) EnforceOverlayMountPolicy(_ string, _ []string) error {
-	return ErrNotImplemented
+func writeEnvRules(builder *strings.Builder, envRules []EnvRuleConfig, indent string) error {
+	_, err := builder.WriteString(fmt.Sprintf("%s\"env_rules\": %s,\n", indent, envRuleArray(envRules).marshalRego()))
+	return err
 }
 
-func (RegoEnforcer) EnforceCreateContainerPolicy(_ string, _, _ []string, _ string) error {
-	return ErrNotImplemented
+func writeLayers(builder *strings.Builder, layers []string, indent string) error {
+	array := (stringArray(layers)).marshalRego()
+	_, err := builder.WriteString(fmt.Sprintf("%s\"layers\": %s,\n", indent, array))
+	return err
 }
 
-func (RegoEnforcer) EnforceWaitMountPointsPolicy(_ string, _ *oci.Spec) error {
-	return ErrNotImplemented
+func (m mountInternal) marshalRego() string {
+	options := stringArray(m.Options).marshalRego()
+	return fmt.Sprintf("{\"destination\": \"%s\", \"options\": %s, \"source\": \"%s\", \"type\": \"%s\"}", m.Destination, options, m.Source, m.Type)
 }
 
-func (RegoEnforcer) EnforceMountPolicy(_, _ string, _ *oci.Spec) error {
-	return ErrNotImplemented
+func writeMounts(builder *strings.Builder, mounts []mountInternal, indent string) error {
+	values := make([]string, len(mounts))
+	for i, mount := range mounts {
+		values[i] = mount.marshalRego()
+	}
+
+	_, err := builder.WriteString(fmt.Sprintf("%s\"mounts\": [%s],\n", indent, strings.Join(values, ",")))
+	return err
 }
 
-func (RegoEnforcer) ExtendDefaultMounts(_ []oci.Mount) error {
-	return ErrNotImplemented
+func writeContainer(builder *strings.Builder, container *securityPolicyContainer, indent string) error {
+	if _, err := builder.WriteString(fmt.Sprintf("%s{\n", indent)); err != nil {
```

Comment:
Fine to leave this as-is, but just so you know, [`(*strings.Builder).WriteString`](https://pkg.go.dev/strings#Builder.WriteString) is guaranteed to always return a `nil` error, so I think we can remove all error handling from the `marshalRego` path.

Reply:
Really the error handling should be that the length of the written string is non-zero.